File: mcserver/download.py
from __future__ import annotations

from pathlib import Path
import sys
from urllib.request import urlopen


def _format_bytes(n: int) -> str:
    value = float(n)
    for unit in ("B", "KB", "MB", "GB"):
        if value < 1024 or unit == "GB":
            if unit == "B":
                return f"{int(value)}{unit}"
            return f"{value:.1f}{unit}"
        value /= 1024
    return f"{int(value)}B"


def download_to(
    url: str,
    dest: Path,
    *,
    chunk_size: int = 1024 * 256,
    label: str = "Downloading",
) -> None:
    dest.parent.mkdir(parents=True, exist_ok=True)
    with urlopen(url) as resp, open(dest, "wb") as f:
        total = resp.headers.get("Content-Length")
        total_bytes = int(total) if total and total.isdigit() else None

        downloaded = 0
        last_pct = -1
        last_bytes_print = 0
        while True:
            chunk = resp.read(chunk_size)
            if not chunk:
                break
            f.write(chunk)
            downloaded += len(chunk)

            if total_bytes:
                pct = int(downloaded * 100 / total_bytes)
                if pct != last_pct and (pct % 2 == 0 or pct == 100):
                    sys.stderr.write(
                        f"\r{label}: {pct:3d}% ({_format_bytes(downloaded)} / {_format_bytes(total_bytes)})"
                    )
                    sys.stderr.flush()
                    last_pct = pct
            else:
                if downloaded - last_bytes_print >= 5 * 1024 * 1024:
                    sys.stderr.write(f"\r{label}: {_format_bytes(downloaded)}")
                    sys.stderr.flush()
                    last_bytes_print = downloaded

    sys.stderr.write("\n")

File: internal/download/download.go
package download

import (
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
)

// DownloadTo downloads a URL to a local file path with progress output to stderr.
func DownloadTo(url, dest, label string) error {
	if err := os.MkdirAll(filepath.Dir(dest), 0o755); err != nil {
		return fmt.Errorf("creating directory: %w", err)
	}

	resp, err := http.Get(url)
	if err != nil {
		return fmt.Errorf("fetching %s: %w", url, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("download failed: HTTP %d", resp.StatusCode)
	}

	f, err := os.Create(dest)
	if err != nil {
		return fmt.Errorf("creating file: %w", err)
	}
	defer f.Close()

	var totalBytes int64
	if cl := resp.Header.Get("Content-Length"); cl != "" {
		totalBytes, _ = strconv.ParseInt(cl, 10, 64)
	}

	buf := make([]byte, 256*1024)
	var downloaded int64
	lastPct := -1

	for {
		n, readErr := resp.Body.Read(buf)
		if n > 0 {
			if _, writeErr := f.Write(buf[:n]); writeErr != nil {
				return fmt.Errorf("writing file: %w", writeErr)
			}
			downloaded += int64(n)

			if totalBytes > 0 {
				pct := int(downloaded * 100 / totalBytes)
				if pct != lastPct && (pct%2 == 0 || pct == 100) {
					fmt.Fprintf(os.Stderr, "\r%s: %3d%% (%s / %s)", label, pct, formatBytes(downloaded), formatBytes(totalBytes))
					lastPct = pct
				}
			}
		}
		if readErr == io.EOF {
			break
		}
		if readErr != nil {
			return fmt.Errorf("reading response: %w", readErr)
		}
	}
	fmt.Fprintln(os.Stderr)
	return nil
}

func formatBytes(n int64) string {
	v := float64(n)
	for _, unit := range []string{"B", "KB", "MB", "GB"} {
		if v < 1024 || unit == "GB" {
			if unit == "B" {
				return fmt.Sprintf("%d%s", int(v), unit)
			}
			return fmt.Sprintf("%.1f%s", v, unit)
		}
		v /= 1024
	}
	return fmt.Sprintf("%dB", int(v))
}
